ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+		Some(true)
+	}
+
+	/// Returns if we are ready to seal, the real sealing (signing extra_data) is actually done in `on_seal_block()`.
+	fn generate_seal(&self, block: &ExecutedBlock, parent: &Header) -> Seal {
+		let null_seal = vec!(encode(&NULL_MIXHASH.to_vec()), encode(&NULL_NONCE.to_vec()));
+
+		trace!(target: "engine", "tried to generate seal");
+
+		if block.header.number() == 0 {
+			trace!(target: "engine", "attempted to seal genesis block");
+			return Seal::None;
+		}
+
+		// if sealing period is 0, and not an checkpoint block, refuse to seal
+		if self.period == 0 {
+			if block.transactions.is_empty() && block.header.number() % self.epoch_length != 0 {
+				return Seal::None;
+			}
+			return Seal::Regular(null_seal);
+		}
+
+		// Check we actually have authority to seal.
+		if let Some(author) = self.signer.read().as_ref().map(|x| x.address()) {
+			// ensure the voting state exists
+			match self.state(&parent) {
+				Err(e) => {
+					warn!(target: "engine", "generate_seal: can't get parent state(number: {}, hash: {}): {} ",
+						  parent.number(), parent.hash(), e);
+					return Seal::None;
+				}
+				Ok(state) => {
+					// Are we authorized to seal?
+					if !state.is_authoirzed(&author) {
+						trace!(target: "engine", "generate_seal: Not authorized to sign right now.");
+						// wait for one third of period to try again.
+						thread::sleep(Duration::from_secs(self.period / 3 + 1));
+						return Seal::None;
+					}
+
+					let inturn = state.is_inturn(block.header.number(), &author);
+
+					let now = SystemTime::now();
+
+					let limit = match inturn {
+						true => state.next_timestamp_inturn.unwrap_or(now),
+						false => state.next_timestamp_noturn.unwrap_or(now),
+					};
+
+					// Wait for the right moment.
+					if now < limit {
+						trace!(target: "engine",
+							   "generate_seal: sleeping to sign: inturn: {}, now: {:?}, to: {:?}.",
+							   inturn, now, limit);
+						match limit.duration_since(SystemTime::now()) {
+							Ok(duration) => {
+								thread::sleep(duration);
+							},
+							Err(e) => {
+								warn!(target:"engine", "generate_seal: unable to sleep, err: {}", e);
+								return Seal::None;
+							}
+						}
+					}
+
+					trace!(target: "engine", "generate_seal: seal ready for block {}, txs: {}.",
+						   block.header.number(), block.transactions.len());
+					return Seal::Regular(null_seal);
+				}
+			}
+		}
+		Seal::None
+	}
+
+	fn verify_local_seal(&self, _header: &Header) -> Result<(), Error> { Ok(()) }
+
+	fn verify_block_basic(&self, header: &Header) -> Result<(), Error> {
+		// Largely same as https://github.com/ethereum/go-ethereum/blob/master/consensus/clique/clique.go#L275
+
+		// Ignore genesis block.
+		if header.number() == 0 {
+			return Ok(());
+		}
+
+		// Don't waste time checking blocks from the future
+		{
+			let limit = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs() + self.period;
```

Comment:
What would the default value be in this case?

Reply:
0, so it doesn't make sense

`SystemTime shouldn't overflow unless it is year 2038 on 32-bit platforms` so return Error in this case. It would be an indication on `hardware problems` unless it is 2038 then